source/particle/particle_handler.cc

```diff
@@ -222,6 +261,661 @@ namespace aspect
+          for (unsigned int i=0; i<send_particles[neighbor_id].size(); ++i)
+            {
+              // If no target cells are given, use the iterator information
+              active_cell_it cell;
+              if (send_cells.size() == 0)
+                cell = send_particles[neighbor_id][i]->get_surrounding_cell(*triangulation);
+              else
+                cell = send_cells[neighbor_id][i];
+
+              const CellId::binary_type cellid = cell->id().template to_binary<dim>();
+              memcpy(data, &cellid, cellid_size);
+              data = static_cast<char *>(data) + cellid_size;
+
+              send_particles[neighbor_id][i]->write_data(data);
+              data = store_callback(send_particles[neighbor_id][i],data);
+            }
+          n_send_data[neighbor_id] = reinterpret_cast<std::size_t> (data) - send_offsets[neighbor_id] - reinterpret_cast<std::size_t> (&send_data.front());
+        }
+
+      // Notify other processors how many particles we will send
+      std::vector<MPI_Request> n_requests(2*n_neighbors);
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        MPI_Irecv(&(n_recv_data[i]), 1, MPI_INT, neighbors[i], 0, mpi_communicator, &(n_requests[2*i]));
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        MPI_Isend(&(n_send_data[i]), 1, MPI_INT, neighbors[i], 0, mpi_communicator, &(n_requests[2*i+1]));
+      MPI_Waitall(2*n_neighbors,&n_requests[0],MPI_STATUSES_IGNORE);
+
+      // Determine how many particles and data we will receive
+      unsigned int total_recv_data = 0;
+      for (unsigned int neighbor_id=0; neighbor_id<n_neighbors; ++neighbor_id)
+        {
+          recv_offsets[neighbor_id] = total_recv_data;
+          total_recv_data += n_recv_data[neighbor_id];
+        }
+
+      // Set up the space for the received particle data
+      std::vector<char> recv_data(total_recv_data);
+
+      // Exchange the particle data between domains
+      std::vector<MPI_Request> requests(2*n_neighbors);
+      unsigned int send_ops = 0;
+      unsigned int recv_ops = 0;
+
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        if (n_recv_data[i] > 0)
+          {
+            MPI_Irecv(&(recv_data[recv_offsets[i]]), n_recv_data[i], MPI_CHAR, neighbors[i], 1, mpi_communicator,&(requests[send_ops]));
+            send_ops++;
+          }
+
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        if (n_send_data[i] > 0)
+          {
+            MPI_Isend(&(send_data[send_offsets[i]]), n_send_data[i], MPI_CHAR, neighbors[i], 1, mpi_communicator,&(requests[send_ops+recv_ops]));
+            recv_ops++;
+          }
+      MPI_Waitall(send_ops+recv_ops,&requests[0],MPI_STATUSES_IGNORE);
+
+      // Put the received particles into the domain if they are in the triangulation
+      const void *recv_data_it = static_cast<const void *> (&recv_data.front());
+
+      while (reinterpret_cast<std::size_t> (recv_data_it) - reinterpret_cast<std::size_t> (&recv_data.front()) < total_recv_data)
```

Comment:
could you avoid the `reinterpret_cast` by instead doing something like
```
  while (rec_data_it < static_cast<void*>(&recv_data.back()))
```
or similar?